src/shims/fs.rs

```diff
@@ -257,6 +263,214 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                }
+            } else {
+                if let Some(path) = &this
+                    .machine
+                    .file_handler
+                    .handles
+                    .get(&dirfd)
+                    .map(|h| &h.path)
+                {
+                    // If `dirfd` points to a directory, `pathname` is relative to that directory
+                    if path.is_dir() {
+                        path.join(pathname)
+                    } else {
+                        // If `pathname` is empty and the `AT_EMPTY_PATH` flag is set, we use the path of `dirfd`.
+                        if pathname.as_os_str().is_empty() && (flags & at_empty_path != 0) {
+                            (*path).clone()
+                        } else {
+                            // It is not specified what should happen here.
+                            throw_unsup_format!(
+                                "the file descriptor `{}` and path `{:?}` combination is invalid.",
+                                dirfd,
+                                path
+                            );
+                        }
+                    }
+                } else {
+                    return this.handle_not_found();
+                }
+            }
+        };
+
+        // the `_mask_op` paramter specifies the file information that the caller requested.
+        // However `statx` is allowed to return information that was not requested or to not
+        // return information that was requested. This `mask` represents the information we can
+        // actually provide in any host platform.
+        let mut mask =
+            this.eval_libc("STATX_TYPE")?.to_u32()? | this.eval_libc("STATX_SIZE")?.to_u32()?;
+
+        // If the `AT_SYMLINK_NOFOLLOW` flag is set, we query the file's metadata without following
+        // symbolic links.
+        let metadata = if flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0 {
+            std::fs::symlink_metadata(path)
+        } else {
+            std::fs::metadata(path)
+        };
+
+        let metadata = match metadata {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                return Ok(-1);
+            }
+        };
+
+        let file_type = metadata.file_type();
+
+        // The `mode` field specifies the type of the file and the permissions over the file for
+        // the owner, its group and other users. Given that we can only provide the file type
+        // without using platform specific methods, we only set the bits corresponding to the file
+        // type.
+        let mode = if file_type.is_file() {
+            this.eval_libc("S_IFREG")?
+        } else if file_type.is_dir() {
+            this.eval_libc("S_IFDIR")?
+        } else {
+            this.eval_libc("S_IFLNK")?
+        }
+        .to_u32()? as u16;
+
+        let size = metadata.len();
+
+        let access_sec;
```

Comment:
Please try to dedup these three time-access snippets

Reply:
I'll write a helper function inside `src/shims/time.rs`

Reply:
check https://github.com/rust-lang/miri/pull/1101/commits/3ace9001a5cb3abf3d78a0e28a3457ecc6d60149